Libstapsdt.cs

```diff
@@ -1,76 +1,98 @@
 
-    [DllImport(LibstapsdtLibrary, CallingConvention = CallingConvention.Cdecl)]
-    public static extern void probeFire(ProbePtr probe);
+    [LibraryImport(LibstapsdtLibrary)]
+    [UnmanagedCallConv(CallConvs = [typeof(CallConvCdecl)])]
+    public static partial void probeFire(ProbePtr probe);
 
-    [DllImport(LibstapsdtLibrary, CallingConvention = CallingConvention.Cdecl)]
-    public static extern void probeFire(ProbePtr probe, long arg1);
+    [LibraryImport(LibstapsdtLibrary)]
+    [UnmanagedCallConv(CallConvs = [typeof(CallConvCdecl)])]
+    public static partial void probeFire(ProbePtr probe, long arg1);
 
-    [DllImport(LibstapsdtLibrary, CallingConvention = CallingConvention.Cdecl)]
-    public static extern void probeFire(ProbePtr probe, long arg1, long arg2);
+    [LibraryImport(LibstapsdtLibrary)]
+    [UnmanagedCallConv(CallConvs = [typeof(CallConvCdecl)])]
+    public static partial void probeFire(ProbePtr probe, long arg1, long arg2);
 
-    [DllImport(LibstapsdtLibrary, CallingConvention = CallingConvention.Cdecl)]
-    public static extern void probeFire(ProbePtr probe, long arg1, long arg2, long arg3);
+    [LibraryImport(LibstapsdtLibrary)]
+    [UnmanagedCallConv(CallConvs = [typeof(CallConvCdecl)])]
+    public static partial void probeFire(ProbePtr probe, long arg1, long arg2, long arg3);
 
-    [DllImport(LibstapsdtLibrary, CallingConvention = CallingConvention.Cdecl)]
-    public static extern void probeFire(ProbePtr probe, long arg1, long arg2, long arg3, long arg4);
+    [LibraryImport(LibstapsdtLibrary)]
+    [UnmanagedCallConv(CallConvs = [typeof(CallConvCdecl)])]
+    public static partial void probeFire(ProbePtr probe, long arg1, long arg2, long arg3, long arg4);
 
-    [DllImport(LibstapsdtLibrary, CallingConvention = CallingConvention.Cdecl)]
-    public static extern void probeFire(ProbePtr probe, long arg1, long arg2, long arg3, long arg4, long arg5);
+    [LibraryImport(LibstapsdtLibrary)]
+    [UnmanagedCallConv(CallConvs = [typeof(CallConvCdecl)])]
+    public static partial void probeFire(ProbePtr probe, long arg1, long arg2, long arg3, long arg4, long arg5);
 
-    [DllImport(LibstapsdtLibrary, CallingConvention = CallingConvention.Cdecl)]
-    public static extern void probeFire(ProbePtr probe, long arg1, long arg2, long arg3, long arg4, long arg5, long arg6);
+    [LibraryImport(LibstapsdtLibrary)]
+    [UnmanagedCallConv(CallConvs = [typeof(CallConvCdecl)])]
+    public static partial void probeFire(ProbePtr probe, long arg1, long arg2, long arg3, long arg4, long arg5, long arg6);
 
-    [DllImport(LibstapsdtLibrary, CallingConvention = CallingConvention.Cdecl)]
-    public static extern bool probeIsEnabled(ProbePtr probe);  // return 1 if true, 0 if false
+    [LibraryImport(LibstapsdtLibrary)]
+    [UnmanagedCallConv(CallConvs = [typeof(CallConvCdecl)])]
+    [return: MarshalAs(UnmanagedType.Bool)]
+    public static partial bool probeIsEnabled(ProbePtr probe);  // return 1 if true, 0 if false
```

Comment:
The [description of Bool](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.unmanagedtype?view=net-8.0#fields) looks a bit suspicious: 

> A 4-byte Boolean value (true != 0, false = 0). This is the **Win32 BOOL type**

Generally speaking, `int` isn't guaranteed to be 4-byte. Perhaps it's easier to just return `int`. What do you think?

```suggestion
    public static partial int probeIsEnabled(ProbePtr probe);  // return 1 if true, 0 if false
```

Reply:
`int` in C# is aliased to `System.Int32` so it's always guaranteed to be 32 bits wide, like `BOOL` from Win32. It doesn't change much except interpretation. I agree that it can be confusing, but if you look at the code generated for marshaling:

```c#
[global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Interop.LibraryImportGenerator", "8.0.10.36612")]
[global::System.Runtime.CompilerServices.SkipLocalsInitAttribute]
public static partial bool probeIsEnabled(nint probe)
{
    bool __retVal;
    int __retVal_native;
    {
        __retVal_native = __PInvoke(probe);
    }

    // Unmarshal - Convert native data to managed data.
    __retVal = __retVal_native != 0;
    return __retVal;
    // Local P/Invoke
    [global::System.Runtime.InteropServices.DllImportAttribute("libstapsdt.so.0", EntryPoint = "probeIsEnabled", ExactSpelling = true)]
    [global::System.Runtime.InteropServices.UnmanagedCallConvAttribute(CallConvs = new global::System.Type[] { typeof(global::System.Runtime.CompilerServices.CallConvCdecl) })]
    static extern unsafe int __PInvoke(nint __probe_native);
}
```

the `bool` is changed to an `int`:

```c#
static extern unsafe int __PInvoke(nint __probe_native);
```

In any case, I think it's very early days and we'll need to review the signatures with more thought and care when it gets time to stabilise and think about various CPU architectures.

Reply:
I mean that the `int` on the side of the C library is not necessarily 4 bytes. But if I understand correctly, this is extremely unlikely, and there's no apparent way to account for this.

The current generated code looks great.

Reply:
> I mean that the `int` on the side of the C library is not necessarily 4 bytes.

`int` in C is implementation-specific so details are muddier than that…`int` in C _on_ Linux x64, which is the runtime platform fixed with PR #2, is 32 bits wide by definition of the [LP64 data model] and therefore compatible with either `int` in C# or Win32 `BOOL`.

[LP64 data model]: https://en.wikipedia.org/wiki/64-bit_computing#64-bit_data_models

If you look at the compiled code for `probeIsEnabled`:

```asm
0000000000003513 <probeIsEnabled>:
    3513:       f3 0f 1e fa             endbr64
    3517:       55                      push   %rbp
    3518:       48 89 e5                mov    %rsp,%rbp
    351b:       48 89 7d f8             mov    %rdi,-0x8(%rbp)
    351f:       48 8b 45 f8             mov    -0x8(%rbp),%rax
    3523:       48 8b 40 20             mov    0x20(%rax),%rax
    3527:       48 85 c0                test   %rax,%rax
    352a:       75 07                   jne    3533 <probeIsEnabled+0x20>
    352c:       b8 00 00 00 00          mov    $0x0,%eax
    3531:       eb 26                   jmp    3559 <probeIsEnabled+0x46>
    3533:       48 8b 45 f8             mov    -0x8(%rbp),%rax
    3537:       48 8b 40 20             mov    0x20(%rax),%rax
    353b:       0f b6 00                movzbl (%rax),%eax
    353e:       0f be c0                movsbl %al,%eax
    3541:       25 90 00 00 00          and    $0x90,%eax
    3546:       3d 90 00 00 00          cmp    $0x90,%eax
    354b:       75 07                   jne    3554 <probeIsEnabled+0x41>
    354d:       b8 00 00 00 00          mov    $0x0,%eax
    3552:       eb 05                   jmp    3559 <probeIsEnabled+0x46>
    3554:       b8 01 00 00 00          mov    $0x1,%eax
    3559:       5d                      pop    %rbp
    355a:       c3                      ret
```

you can see that the result of the function is being fully loaded into the EAX register each time (at offsets 0x352c, 0x354d and 0x3554) and the EAX register is 32 bits wide (thus proving that the C compiler treated `int` as 32-bit data). If it was `RAX`, it would have been 64 bits wide. You can also double-check this with a simple program on 64-bit Linux:

```c
#include <stdio.h>

void main() {
    printf("sizeof(int) = %d\n", sizeof(int)); // prints: sizeof(int) = 4
}
```

What I was doing here was getting the marshalling layer to treat the _32-bit_ `int` in C (assuming LP64) as the 32-bit `BOOL` from Win32 and convert it to a C# `bool` where _true_ is `!0` and _false_ is `0`. This is all compatible, but I can understand that it can be confusing for someone to see `int` on the C side and `bool` on the C# side with the marshaling layer being told that it's a Win32 `BOOL`. To be honest, when you read the [`UnmanagedType`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.unmanagedtype) documentation for `Bool`, it says:

> A 4-byte Boolean value (`true` != 0, `false` = 0). This is the Win32 BOOL type.

So it's really just a 4-byte Boolean first. The Win32 `BOOL` happens to be sharing the same definition and so it should really read:

> A 4-byte Boolean value (`true` != 0, `false` = 0). This is _also_ the Win32 BOOL type.

Better yet:

> A 4-byte Boolean value (`true` != 0, `false` = 0). This is _also_ the Win32 BOOL type or an in `int` in C on the LP64 data model being used as a Boolean.

If you look at the `UnmanagedType.I1` documentation, it specifically mentions C-style `bool`:

> A 1-byte signed integer. You can use this member to transform a Boolean value into a 1-byte, C-style `bool` (`true` = 1, `false` = 0).

[C99]: https://en.wikipedia.org/wiki/C99

Reply:
Do you want to update MSFT documentation? It would really be clearer :)

I didn't know about LP64, interesting to learn how it's standardized and can be assumed for a Linux system. Also interesting how the compiler actually puts a "1" as a return value rather than "just something nonzero" which could likely save some cycles.